File: 0143-reorder-list/0143-reorder-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {void} Do not return anything, modify head in-place instead.
 */
var reorderList = function(head) {
  // Find middle of linked list
  let slow = head, fast = head.next;
  while (fast && fast.next) {
    slow = slow.next;
    fast = fast.next.next;
  }
  
  let secondHalf = slow.next;
  slow.next = null;
  
  let prev = null;
  while (secondHalf) {
    let temp = secondHalf.next;
    secondHalf.next = prev;
    prev = secondHalf;
    secondHalf = temp;
  }
  
  // Merge both halfs the lists
  let firstHalf = head;
  secondHalf = prev;
  
  while (secondHalf) {
    let [temp1, temp2] = [firstHalf.next, secondHalf.next];
    firstHalf.next = secondHalf;
    secondHalf.next = temp1;
    firstHalf = temp1;
    secondHalf = temp2;
  }
  
};

File: 0143-reorder-list/0143-reorder-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
      # Split list into two halfs by finding middle
      slow, fast = head, head.next
      while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
      
      # Second half starts at the middle node's next
      second_half = slow.next
      # The next of last node in first half should be None
      slow.next = None
      
      # Reverse second half
      prev = None
      while second_half:
        temp = second_half.next
        second_half.next = prev
        prev = second_half
        second_half = temp
      
      # Second half's new head is at prev b/c second half is None 
      second_half = prev
      first_half = head
      
      while second_half:
        temp1, temp2 = first_half.next, second_half.next
        first_half.next = second_half
        second_half.next = temp1
        first_half, second_half = temp1, temp2
        